### PR TITLE
Fixed bug in FixInvalidSsaTag + Refact + TEST

### PR DESCRIPTION
**Pull request by ivandrofly**

### CODE REVIEW

File: src/Core/HtmlUtil.cs [hunk 1]
```diff
@@ -9,7 +9,7 @@ namespace Nikse.SubtitleEdit.Core
     /// <summary>
     /// HTML specific string manipulations.
     /// </summary>
-    internal static class HtmlUtil
+    public static class HtmlUtil
```
Comment (ivandrofly) [inline src/Core/HtmlUtil.cs]: Not visible to unitest

File: src/Core/HtmlUtil.cs [hunk 2]
```diff
@@ -380,5 +377,253 @@ public static string RemoveHtmlTags(string s, bool alsoSsaTags)
 
             return s;
         }
+
+        public static string FixInvalidItalicTags(string text)
```
Comment (ivandrofly) [inline src/Core/HtmlUtil.cs]: Moved to Htmlutil

File: src/Forms/Main.cs [hunk 3]
```diff
@@ -6505,18 +6505,6 @@ private void MoveNextPrevious(int firstSelectedIndex)
                 SubtitleListview1.SelectIndexAndEnsureVisible(firstSelectedIndex);
         }
 
-        private static string RemoveSsaStyle(string text)
```
Comment (ivandrofly) [inline src/Forms/Main.cs]: Already defined

File: src/Logic/Utilities.cs [hunk 4]
```diff
@@ -807,12 +807,11 @@ private static string ReInsertHtmlTags(string s, Dictionary<int, string> htmlTag
 
         public static string UnbreakLine(string text)
         {
-            if (text.Contains(Environment.NewLine))
+            if (text.Contains(Environment.NewLine, StringComparison.Ordinal))
             {
-                string newText = text.Replace(Environment.NewLine, " ");
```
Comment (ivandrofly) [inline src/Logic/Utilities.cs]: No new var needed

File: src/Logic/Utilities.cs [hunk 5]
```diff
@@ -841,22 +840,15 @@ public static void InitializeSubtitleFont(Control control)
 
         public static string RemoveSsaTags(string s)
         {
-            int k = s.IndexOf('{');
-            while (k >= 0)
+            if (s == null || !s.Contains("\\a", StringComparison.Ordinal))
```
Comment (ivandrofly) [inline src/Logic/Utilities.cs]: More safe

File: src/Logic/Utilities.cs [hunk 6]
```diff
@@ -3473,5 +3220,35 @@ public static int GetNumberOfLines(string text)
...
+            var pre = string.Empty;
+            var idx = text.IndexOf("{\\a", StringComparison.Ordinal);
+            var firstTime = idx;
+            while (idx >= 0)
+            {
+                var endIdx = text.IndexOf('}', idx + 2);
+                if (endIdx < idx)
+                    break;
+                if (firstTime == idx)
+                    pre = text.Substring(idx, endIdx - idx + 1);
```
Comment (ivandrofly) [inline src/Logic/Utilities.cs]: Most player only looks for one Ssa tag all the others will be ignore, so move it at the beginning

File: src/Test/UtilitiesTest.cs [hunk 7]
```diff
@@ -391,5 +399,26 @@ public void RemoveLineBreaks5()
             Assert.AreEqual(result, "<i>Foobar</i>");
         }
 
+        [TestMethod]
+        [DeploymentItem("SubtitleEdit.exe")]
+        public void FixInvalidSsaTagsTest1()
+        {
+            var current = "Who is {\\an7}next?";
```
Comment (ivandrofly) [inline src/Test/UtilitiesTest.cs]: Have seen something like this before

File: src/Core/HtmlUtil.cs (hunk 1 above)
Comment (niksedk) [inline src/Core/HtmlUtil.cs]: it should be...

File: src/Core/HtmlUtil.cs (hunk 2 above)
Comment (niksedk) [inline src/Core/HtmlUtil.cs]: :)

File: src/Logic/Utilities.cs [hunk 8]
```diff
@@ -3473,5 +3220,35 @@ public static int GetNumberOfLines(string text)
...
         }
 
+        public static string FixInvalidSsaTags(string text)
+        {
+            if (text == null || text.Length <= 6)
+                return text;
+
+            text = text.Replace("{ \\a", "{\\a");
+            text = text.Replace("{\\a ", "{\\a");
+            text = text.Replace("{\\an ", "{\\an");
```
Comment (niksedk) [inline src/Logic/Utilities.cs]: Try to avoid the \alpha tag...

Comment (niksedk) [inline src/Logic/Utilities.cs]: (Could be missing more places...)

File: src/Core/HtmlUtil.cs (hunk 1 above)
Comment (ivandrofly) [inline src/Core/HtmlUtil.cs]: yeah truth... thought would be same for plugin's test :wink:

File: src/Logic/Utilities.cs (hunk 8 above)
Comment (ivandrofly) [inline src/Logic/Utilities.cs]: :+1: all done in https://github.com/SubtitleEdit/subtitleedit/pull/555

**Pull request closed without merging**